File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.java
import java.util.*;
import java.io.*;

class Solution {
    public int solution(int cacheSize, String[] cities) {
        int answer = 0;
        
        Queue<String> q = new LinkedList<>();
        
        // 대소문자 구분 없이 처리
        for (int i=0; i < cities.length; i++) {
            cities[i] = cities[i].toLowerCase();
        }
        
        //System.out.println(Arrays.toString(cities));
        
        for (String c : cities) {
            
            // cache hit
            if (q.contains(c)) {
                q.remove(c);
                q.add(c);
                answer += 1;
            }
            
            // cache miss
            else {
                answer += 5;
                if (q.size() >= cacheSize) {
                    q.poll();
                    q.add(c);
                    
                    if (cacheSize == 0) {
                        q.poll();
                    }
                }
                
                else {
                    q.add(c);
                }
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.py
from collections import deque

def solution(cacheSize, cities):
    answer = 0
    
    for i in range(len(cities)) :
        cities[i] = cities[i].lower()
    
    ##print(cities)
    q = deque()
    
    for city in cities :
        
        ## cacheSize == 0인 경우
        if cacheSize == 0 :
            answer += 5
            continue
        
        ## 큐가 아예 비어있을 경우
        if len(q) == 0 :
            q.append(city)
            answer += 5
            continue
        
        ## cache hit
        if city in q :
            q.remove(city)
            q.append(city)
            answer += 1
        
        ## cache miss
        else :
            ## cacheSize에 도달한 경우
            if  cacheSize == len(q) :
                q.popleft()
            
            q.append(city)
            answer += 5
    
    return answer